packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -3,19 +3,114 @@
+///
+/// The full error message should include at least the following elements:
+/// * Claim: Explanation of the assertion failure or contract violation.
+/// * Grounds: Facts about the user's code that led to the error.
+/// * Warranty: Connections between the grounds and the claim.
+///
+/// Error messages may include references to objects relevant to the error using
+/// string interpolation and those objects will be directly available to
+/// debugging tools when error messages are generating as part of a debug builds.
+/// Other builds will see the same string error message but will not be able to
+/// extract the contents of the object.
+class ErrorMessage extends _ErrorDiagnostic {
+  // This constructor provides a reliable hook for a kernel transformer to find
+  // error messages that need to be rewritten to include object references.
+  //
+  // The message will display with the same text regardless of whether the
+  // kernel transformer is used but without the kernel transformer, debugging
+  // tools are unable to provide interactive displays of objects referenced by
+  // the string.
+  /// In debug builds, a kernel transformer rewrites calls to the default
+  /// constructor into calls to this constructor.
+  ///
+  /// of this class as follows so that the objects referenced in the
+  /// string literal can be captured by debuggers.
+  /// ```dart
+  /// ErrorMessage('Element $element must be color $color')
+  /// ```
+  /// Desugars to:
+  /// ```dart
+  /// ErrorMessage._structured(<Object>['Element ', element, ' must be ', color]);
+  /// ```
+  ///
+  /// Slightly more complext case:
+  /// ```dart
+  /// ErrorMessage('Element ${element.runtimeType} must be must be $color')
+  /// ```
+  /// Desugars to:
+  ///```dart
+  /// ErrorMessage._structured(<Object>[
+  ///   'Element ',
+  ///    DiagnosticsProperty(null, element, description: element.runtimeType?.toString()),
+  ///    ' must be ',
+  ///    color,
+  /// ]);
+  /// ```
+  ErrorMessage(String message) : super._fromParts([message], level: DiagnosticLevel.summary);
+
+  // Constructor that calls to the default constructor are rewriten to use in
+  // debug mode using a kernel transformer.
+  ErrorMessage._fromParts(List<Object> messageParts) : super._fromParts(messageParts, level: DiagnosticLevel.summary);
+}
+
+/// An [ErrorHint] is a suggestion for resolving the error.
+///
+/// Urls may be be included in the hint to reference external material.
+class ErrorHint extends _ErrorDiagnostic {
+  // A lint enforces that this constructor can only be called with a string
+  // literal to match the limitations of the Dart Kernel transformer that
+  // optionally extracts out objects referenced using string interpolation in
+  // the message passed in.
+  ErrorHint(String message) : super._fromParts([message], level:DiagnosticLevel.hint);
+
+  ErrorHint._structured(List<Object> messageParts) : super._fromParts(messageParts, level:DiagnosticLevel.hint);
```

Comment:
assuming `ErrorHint._structured` is used by the kernel transformer, add a comment to that effect

Reply:
Indeed it is. done.